### PR TITLE
finagle-thrift: Update serverFromIface to try new style constructor.

### PR DESCRIPTION
**Pull request by zfy0701**
Problem:
In https://github.com/twitter/scrooge/pull/211, an new constructor with service name as
paramter is added.

Solution:
Try to find the new style constructor but fallback to old one if not found.

### CODE REVIEW

File: finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala [hunk 1]
```diff
@@ -145,7 +146,8 @@ private[twitter] object ThriftUtil {
     impl: AnyRef,
     protocolFactory: TProtocolFactory,
     stats: StatsReceiver,
-    maxThriftBufferSize: Int
+    maxThriftBufferSize: Int,
+    label: String
```
Comment (mosesn) [inline finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala]: hmm, we don't want to break binary backcompat.  can you overload with the old method too?

File: finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala [hunk 2]
```diff
@@ -155,14 +157,25 @@ private[twitter] object ThriftUtil {
...
-      for {
+      (for {
         baseName   <- findRootWithSuffix(iface.getName, "$FutureIface") orElse
           Some(iface.getName)
         serviceCls <- findClass[BinaryService](baseName + "$FinagleService") orElse
           findClass[BinaryService](baseName + "$FinagledService")
         baseClass  <- findClass1(baseName)
-        cons       <- findConstructor(serviceCls, baseClass, classOf[TProtocolFactory], classOf[StatsReceiver], Integer.TYPE)
-      } yield cons.newInstance(impl, protocolFactory, stats, Int.box(maxThriftBufferSize))
+      } yield {
```
Comment (mosesn) [inline finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala]: oh god.  thanks for doing this, this is crazy.  could you add a comment as to why we need this?

**Comment by mosesn** [comment]
LGTM except for comment, will try to get more eyes on this.

**Comment by blackicewei** [comment]
LGTM.  @zfy0701, can you resolve Moses's comment? I will pull it in afterwards.

**Comment by zfy0701** [comment]
Added comments. Thank you

**Comment by vkostyukov** [comment]
Thanks @zfy0701! I'm pulling this internally.

**Comment by mosesn** [comment]
:+1:

**Comment by vkostyukov** [comment]
Hey @zfy0701!

Any chance we can test this? I run this PR throght the sandbox and I can see a couple of failed tests internally. Basically, it crashes on `cons => cons.newInstance(newArgs: _*)` with "Wrong number of arguments". 

Do you think we can test this for both cases: w/o new constructor and w/ new constructor?

**Comment by zfy0701** [comment]
@vkostyukov sorry I just realized I did something wrong (I should cast the newArgs as a Seq[Object]) and I probably didn't run the right set of test to check it.

I could do manual check on both case, but having automatic test on both version might be hard (I'm really really new to  finagle/scrooge repository).

Also it would be great if you could let me know which test is failing.

**Comment by zfy0701** [comment]
@vkostyukov I found the cause is that I indeed pass one more extra argument for oldArgs, and I'm able to reproduce the problem by running finagle tests, after my this fix, the problem should be no more, could you try pull it again?

**Comment by vkostyukov** [comment]
Thanks @zfy0701! I'm happy to pull this again! Will update this PR once it passes or not passes the internal tests.

**Comment by mosesn** [comment]
This was merged internally, thanks for your patience!  I'll update this ticket again when we publish it to the `develop` branch.

**Comment by vkostyukov** [comment]
This made it to the develop branch: https://github.com/twitter/finagle/commit/d2453314d0a9411a20c82d83cd9eebbfc7fd3288

**Pull request closed without merging**